cpp/log/cert.cc

```diff
@@ -546,12 +588,337 @@ Cert::Status Cert::ExtensionStructure(int extension_nid,
+
+  // First. Check all the Subject Alt Name extension records. Any that are of
+  // type DNS must pass validation if they are attempting to redact labels
+  STACK_OF(GENERAL_NAME)* subject_alt_names =
+      static_cast<STACK_OF(GENERAL_NAME)*>(
+          X509_get_ext_d2i(x509_, NID_subject_alt_name, NULL, NULL));
+
+  if (subject_alt_names) {
+    const int subject_alt_name_count = sk_GENERAL_NAME_num(subject_alt_names);
+
+    for (int i = 0; i < subject_alt_name_count; ++i) {
+      GENERAL_NAME* const name(sk_GENERAL_NAME_value(subject_alt_names, i));
+
+      Cert::Status name_status;
+
+      if (name->type == GEN_DNS) {
+        const string dns_name = ASN1ToStringAndCheckForNulls(name->d.dNSName,
+                                                             "DNS name",
+                                                             &name_status);
+
+        if (name_status != TRUE) {
+          sk_GENERAL_NAME_free(subject_alt_names);
+          *status = name_status;
+          return true;
+        }
+
+        dns_alt_names.push_back(dns_name);
+
+        if (IsRedactedHost(dns_name)) {
+          if (!IsValidRedactedHost(dns_name)) {
+            LOG(WARNING) << "Invalid redacted host: " << dns_name;
+            sk_GENERAL_NAME_free(subject_alt_names);
+            *status = Cert::FALSE;
+            return true;
+          }
+
+          redacted_name_count++;
+        }
+      }
+    }
+
+    sk_GENERAL_NAME_free(subject_alt_names);
+  }
+
+  // The next stage of validation is that if the subject name CN exists it
+  // must match the first DNS id and have the same labels redacted
+  // TODO: Confirm it's valid to not have a CN.
+  X509_NAME* const name(X509_get_subject_name(x509_));
+
+  if (!name) {
+    LOG(ERROR) << "Missing X509 subject name";
+    *status = Cert::ERROR;
+    return true;
+  }
+
+  int name_pos = -1;
+
+  name_pos = X509_NAME_get_index_by_NID(name, NID_commonName, name_pos);
+
+  if (name_pos >= 0) {
+    X509_NAME_ENTRY* const name_entry(X509_NAME_get_entry(name, name_pos));
+
+    if (name_entry) {
+      ASN1_STRING * const subject_name_asn1(
+          X509_NAME_ENTRY_get_data(name_entry));
+
+      if (!subject_name_asn1) {
+        LOG(WARNING) << "Missing subject name";
+        // TODO: Check this is correct behaviour. Is it OK to not have
+        // a subject?
+      } else {
+        Cert::Status cn_status;
+        common_name = ASN1ToStringAndCheckForNulls(subject_name_asn1,
+                                                   "CN", &cn_status);
+
+        if (cn_status != TRUE) {
+          *status = cn_status;
+          return true;
+        }
+      }
+    }
+  }
+
+  // If both a subject CN and DNS ids are present in the cert then the
+  // first DNS id must exactly match the CN
+  if (!dns_alt_names.empty() && common_name.length() > 0) {
+    if (dns_alt_names[0] != common_name) {
+      LOG(WARNING) << "CN " << common_name << " does not match DNS.0 "
+          << dns_alt_names[0];
```

Comment:
Style: the `<<` should be indented to go along the first ones on the previous line.